### ISSUE TITLE
Encode complex types only once

### ISSUE THREAD

**Issue by robert-e-davidson3** [onflow/cadence#1973]
Right now, the complex types are encoded in-line with their values. If two values have the same type, the type is encoded twice.

The complex types should only be encoded once. There are two approaches; part of this ticket is to decide on an approach:
1. Encode the types in a header. Each instance of the type will be replaced with a pointer to the type in the header. A pointer in this case is the byte offset in the header.
2. Continue to encode types in-line. The first type will be encoded as is done today but subsequent types will be encoded as pointers to that in-lined type.

The first approach has the advantage of providing a header that can be partially or totally omitted [if the client already knows the types](https://github.com/onflow/cadence/issues/1714 ). (It's only partial when encoding an abstract type because a concrete type is needed for decoding.)

The second approach is a little smaller and faster.

It may be desirable to support both approaches. This would be done by encoding that as a feature flag in the first byte of the encoding, where the codec type and version is indicated.

**Comment by j1010001** [comment]
Closing in favor of redesigned solution captured in: https://github.com/onflow/cadence/issues/2157
